File: leetcode_python/32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        dp = [0] * len(s)
        for i in range(1, len(s)):
            if s[i] == '(': continue
            if s[i - 1] == '(':
                dp[i] = dp[i - 2] + 2
            elif i > dp[i - 1] and s[i - dp[i - 1] - 1] == '(':
                dp[i] = dp[i - 1] + 2 + dp[i - dp[i - 1] - 2]
        return max(dp) if s else 0

File: leetcode_java/32. Longest Valid Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int[] dp = new int[s.length()];
        int res = 0;
        for(int i = 1; i < s.length(); i++) {
            if(s.charAt(i) == '(') continue;
            if(s.charAt(i - 1) == '(')
                dp[i] = 2 + (i > 1 ? dp[i - 2] : 0);
            else if(i > dp[i - 1] && s.charAt(i - dp[i - 1] - 1) == '(')
                dp[i] = dp[i - 1] + 2 + (i - dp[i - 1] >= 2 ? dp[i - dp[i - 1] - 2] : 0);
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}